Fix undefined order of functions when generating code for index expressions (#14374)
ethereum/solidity: Changelog.md

 * Allow qualified access to events from other contracts.

Compiler Features:

>  * Code Generator: Fix not entirely deterministic order of functions in unoptimized Yul output. The choice of C++ compiler in some cases would result in different (but equivalent) bytecode. Official binaries are affected.

Reviewer: ```suggestion
 * Code Generator: Fix not entirely deterministic order of functions in unoptimized Yul output. The choice of C++ compiler in some cases would result in different (but equivalent) bytecode.
```

I'd myself skip that part - it doesn't really belong into the changelog which describes code changes and not means of distributing. But fine either way.

Author: I want to make it clear that emscripten and native binaries produce different bytecode. You wanted me to make it short so I resorted to this.

How about this instead?
```suggestion
 * Code Generator: Fix not entirely deterministic order of functions in unoptimized Yul output. The choice of C++ compiler in some cases would result in different (but equivalent) bytecode (especially native binaries vs emscripten binaries).
```

Reviewer: That version sounds fine to me.

Author: ok then. Changed to that version.

>  * Code Generator: Fix not entirely deterministic order of functions in unoptimized Yul output. The choice of C++ compiler in some cases would result in different (but equivalent) bytecode (especially from native binaries vs emscripten binaries)

Reviewer 2: ```suggestion
 * Code Generator: Fix not entirely deterministic order of functions in unoptimized Yul output. The choice of C++ compiler in some cases would result in different (but equivalent) bytecode, especially from native binaries vs emscripten binaries.
```

Also, shouldn't this be in the  `Bugfixes` section?

Reviewer 2: #14355 seems to have a similar entry, but with an improved version of the text. Just noting this so we don't forget or overwrite that.

Author: > Also, shouldn't this be in the `Bugfixes` section?

Good catch. Of course it should :) Looks like I did not pay enough attention when I was moving it :)

> #14355 seems to have a similar entry, but with an improved version of the text. Just noting this so we don't forget or overwrite that.

Yeah, that's because it's based on this PR. No worries, one reason for keeping it as a draft is to avoid someone merging it accidentally without first merging this PR.

Author: Fixed.

>  * Code Generator: Fix not entirely deterministic order of functions in unoptimized Yul output (i.e. ``--ir`` on the CLI or ``ir`` in Standard JSON). The order would depend on the C++ compiler used to build the Solidity compiler binary, and resulted in GCC builds generating slightly different (but equivalent) bytecode than Clang builds.

Reviewer: We should try to keep changelog entries more concise.

Author: I intentionally made this longer because I wanted to mention the consequences of this bug. All chain explorers offering source verification are affected and need to apply a workaround so this should not be easily confused with some unimportant internal detail.

I tried to omit some detail while still keeping the important bits. Check how you like it now.

Bugfixes:

>  * Code Generator: Fix not entirely deterministic order of functions in unoptimized Yul output. The choice of C++ compiler in some cases would result in different (but equivalent) bytecode (especially from native binaries vs emscripten binaries)

Reviewer 3: ```suggestion
 * Code Generator: Fix not entirely deterministic order of functions in unoptimized Yul output. The choice of C++ compiler in some cases would result in different (but equivalent) bytecode (especially from native binaries vs emscripten binaries).
```

Author: Well, we can still fix this when we sort the changelog before release. If this was a bigger mistake I'd push a new PR, but for a dot it's not worth it :)
